libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
+/// # Safety
+///
+/// - `file`, if non-null, must be an open file handle obtained from [`gzopen`] or [`gzdopen`].
+/// - `buf` must point to at least `len` bytes of readable memory.
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzwrite))]
+pub unsafe extern "C-unwind" fn gzwrite(file: gzFile, buf: *const c_void, len: c_uint) -> c_int {
+    let Some(state) = (unsafe { file.cast::<GzState>().as_mut() }) else {
+        return 0;
+    };
+
+    // Check that we're writing and that there's no error.
+    if state.mode != GZ_WRITE || state.err != Z_OK {
+        return 0;
+    }
+
+    // Check that the requested number of bytes can be represented by the return type.
+    if c_int::try_from(len).is_err() {
+        // Safety: we confirmed above that state is valid.
+        unsafe {
+            gz_error(
+                state,
+                Some((Z_DATA_ERROR, "requested length does not fit in int")),
+            )
+        };
+        return 0;
+    }
+
+    // Safety: We validated state above, and the caller is responsible for ensuring
+    // that buf points to at least len bytes of readable memory.
+    unsafe { gz_write(state, buf, len) }
+}
+
+// Internal implementation of `gzwrite`.
+//
+// # Returns
+// - The number of uncompress bytes written, on success.
+// - Or 0 in case of error.
+//
+// # Safety
+//
+/// - `state` must have been properly initialized, e.g. by [`gzopen_help`].
+/// - `buf` must point to at least `len` bytes of readable memory.
+/// - The caller must ensure that `len` fits in a `c_int`.
+unsafe fn gz_write(state: &mut GzState, mut buf: *const c_void, mut len: c_uint) -> c_int {
+    // If len is zero, avoid unnecessary operations.
+    if len == 0 {
+        return 0;
+    }
+
+    // Allocate memory if this is the first time through.
+    if state.input.is_null() && gz_init(state).is_err() {
+        return 0;
+    }
+
+    /* FIXME uncomment once seek support is implemented
+    if state.seek {
+        state.seek = 0;
+        if gz_zero(state, state.skip) == -1 {
+            return 0;
+        }
+    }
+     */
+
+    let put = len as c_int;
+
+    // For small len, copy to input buffer, otherwise compress directly.
+    if (len as usize) < state.in_size {
+        // Copy to input buffer, compress when full.
+        loop {
+            if state.stream.avail_in == 0 {
+                state.stream.next_in = state.input;
+            }
+            let have = unsafe {
+                state
+                    .stream
+                    .next_in
+                    .add(state.stream.avail_in as usize)
+                    .offset_from(state.input)
+            } as usize;
+            let copy = cmp::min(state.in_size.saturating_sub(have), len as usize);
+            // Safety: The caller is responsible for ensuring that buf points to at least len readable
+            // bytes, and copy is <= len.
+            unsafe { ptr::copy(buf, state.input.add(have).cast::<c_void>(), copy) };
```

Comment:
`buf` is only readable memory (per the safety comment), so this copy should/could be non-overlapping then right?

Reply:
The C API exposes `state.next` as part of `gzFile_s`, so it's possible for C code using this library to do something like:
```
void* buf = file->next;
unsigned len = file->have;
gzwrite(file, buf, len);
```

I think `next` can only ever point into `state.output` and not `state.input`, so even a call like that will lead to a non-overlapping copy here, but I need to look through all the code paths to make sure.